### ISSUE TITLE
3.5 corrupts completely all animated sprites imported as atlas textures/regions from 3.44

### ISSUE THREAD

**Issue by chucklepie** [godotengine/godot#70969]
### Godot version

3.5.1 stable

### System information

linux

### Issue description

Find attached 3 projects:

project_344: this is the original project, open it with 3.4.4 and observe the animated sprite is ok in the scene

project_351: this is a copy of project_344, no modification have been made and simply all that happened was project was opened then saved and closed

project_351_fixed: this is a copy of project_344 but the texture on the animated sprite was redone. This is not required to be looked at, simply a reference to show godot 3.5 has done something bad on converting regions.

The original 'project_344' and 'project_351_fixed' has this entry for a reference to the sprite in the .tscn file:

region = Rect2( 192, 192, 64, 64 )

In the file 351, which is an exact copy of 344 has this entry:

region = Rect2( 768, 1140, 64, 64 )

So when 351 is opened it has read the scene file and for whatever reason has applied an offset to texture region.

This means that a project cannot be upgraded from 344 to 3.5 or 4 and I have no idea how to fix this, meaning mine, and anyone using regions, have broken projects.

If a known fault, can somebody provide guidance on a fix because I am now stuck in limbo, unable to do anything? Or if not/either way, how can I get it to open my scene files without converting...

### Steps to reproduce

open up project_344 and observe the animated sprite is ok

open up proejct_351 and observe the animated sprite is blank and the tscn file has added an offset to the regions for some reason, causing everything to be wrong.

### Minimal reproduction project

[tresbug.zip](https://github.com/godotengine/godot/files/10355750/tresbug.zip)

**Comment by kleonc** [comment]
@chucklepie That's caused by #56795 which e.g. makes your issue #56767 fixed in 3.5+. Quoting myself:
>Just note that for the frames already saved with the incorrect data above issues still hold.

And the frames saved in 3.4 do have incorrect data because the saved frame's region combines all the intermediate AtlasTextures. `Rect2( 768, 1140, 64, 64 )` is just `Rect2( 192, 192, 64, 64 )` shifted by the position of the parent AtlasTexture:

`(192, 192) + (576, 948) == (768, 1140)`
![Godot_v3 5 1-stable_win64_llwkNNF3oY](https://user-images.githubusercontent.com/9283098/210892929-c84c21c1-724d-4f1a-acdc-06e19cf920f0.png)

In 3.5 only the actual subregion (`Rect2( 192, 192, 64, 64 )`) of the AtlasTexture is saved. This means that e.g. if you'd change the source of that AtlasTexture (like in #56767) then the subregion would still be interpreted properly.


So yeah, when it comes to nested AtlasTextures the 3.5 is not compatibile with 3.4 because of that bugfix.

>This means that a project cannot be upgraded from 344 to 3.5 or 4 and I have no idea how to fix this, meaning mine, and anyone using regions, have broken projects.
>
>If a known fault, can somebody provide guidance on a fix because I am now stuck in limbo, unable to do anything? Or if not/either way, how can I get it to open my scene files without converting...

I guess it should be possible to write a script which would find SpriteFrames resources and fix them based on the referred AtlasTextures' data. I could try providing such script. Do you have all SpriteFrames saved as a separate files or some of them in your project are saved as subresources withing the scene file?

**Comment by chucklepie** [comment]
So a bug I raised is fixed, but it then that fix affects more stuff that I've also found, maybe I'm the only person in the Godot universe using atlases/regions for animated sprites ;)

In retrospect it would be nice to have a fix for this, but I guess I can go through every animated sprite I have and modify every single item.

But the changelog does say this, 'We'll update this list with any further regression that the first users of 3.5-stable may report.' 

Could it be updated so anyone else thinking of migrating knows anything using regions will be broken?

**Comment by kleonc** [comment]
@chucklepie Here's a script (likely can be improved) that I think should modify SpriteFrames created in (post-#51030 and before-#56795)(e.g. in 3.4.4) to work properly in (post-#56795)(e.g. 3.5). Some notes:
- it assumes all frames in SpriteFrames referencing another AtlasTexture were created by using the _"Add Frames from Sprite Sheet"_ dialog (in other words it assumes all such frames are affected by this issue),
- when running with `var overwrite_original_files := true` it should be run only once because subsequent runs will keep modifying relevant SpriteFrames; this would break their regions more and more.

So I suggest you to try:
1. Open your 3.4.4 project (of course make sure you have it backuped in case something will go wrong) in 3.5.1.
2. Add the fixing script (create new script and copy paste the code of `SpriteFrames_fixer.gd` from below).
3. Might be a good idea to close all scenes.
4. Run the fixing script (open it in the text editor and run it (Ctrl+Shift+X by default, or through menu: File -> Run)) with line 6 being `var overwrite_original_files := false`.
1. Some fixed files should be created, check the output.
2. Check if they're indeed fixed properly.
3. If everything is good, you could run the fixing script once again, with line 6 changed to `var overwrite_original_files := true` (again, might be a good idea to close all scenes before).
4. If everything is good, get rid of that script to don't ever run it again by mistake.


<details>
<summary>SpriteFrames_fixer.gd</summary>

```gdscript
tool
extends EditorScript


func _run() -> void:
	var overwrite_original_files := false
	fix_sprite_frames_in("res://", overwrite_original_files)
	get_editor_interface().get_resource_filesystem().scan()


func fix_sprite_frames_in(path: String, overwrite_original_files := false) -> void:
	var dir := Directory.new()
	if dir.open(path) != OK:
		print("Couldn't open: %s" % [path])
		return

	var skip_navigational := true
	var skip_hidden := true
	if dir.list_dir_begin(skip_navigational, skip_hidden) != OK:
		print("Couldn't list_dir_begin: %s" % [path])
		return

	var resources_to_fix := []

	var file_name := dir.get_next()
	while file_name != "":
		var file_path := path.plus_file(file_name)
		if dir.current_is_dir():
			fix_sprite_frames_in(file_path, overwrite_original_files)
		elif ResourceLoader.exists(file_path):
			var skip_cache := true
			var resource := ResourceLoader.load(file_path, "", skip_cache)
			if resource is PackedScene or resource is SpriteFrames:
				resources_to_fix.push_back(resource)

		file_name = dir.get_next()

	for resource in resources_to_fix:
		if resource is PackedScene:
			fix_packed_scene_resource_file(resource, overwrite_original_files)
		elif resource is SpriteFrames:
			fix_sprite_frames_resource_file(resource, overwrite_original_files)


func fix_packed_scene_resource_file(scene: PackedScene, overwrite_original_file := false) -> void:
	if scene.resource_path.get_basename().ends_with("_fixed"):
		return # Let's ignore files produced by us.

	var scene_instance := scene.instance()

	var found_sprite_frames: Array = get_sprite_frames_from_node_subtree(scene_instance)

	var anything_fixed := false
	for sprite_frames in found_sprite_frames:
		if try_fix_sprite_frames(sprite_frames):
			anything_fixed = true

	if not anything_fixed:
		scene_instance.free() # Don't leak.
		return # No SpriteFrames needed fixing.

	# Something was fixed, try to resave.
	var new_scene := PackedScene.new()
	var pack_error := new_scene.pack(scene_instance)
	scene_instance.free() # No longer needed, already repacked.

	if pack_error != OK:
		print("Couldn't repack: %s" % [scene.resource_path])
		return

	var new_path := "%s_fixed.%s" % [scene.resource_path.get_basename(), scene.resource_path.get_extension()]
	var save_error := ResourceSaver.save(new_path, new_scene, ResourceSaver.FLAG_CHANGE_PATH)
	if save_error != OK:
		print("Couldn't save: %s" % [new_path])
		return

	print("Fixed scene: %s" % [scene.resource_path])
	print(" saved to: %s" % [new_path])

	if overwrite_original_file:
		var rename_error := Directory.new().rename(new_path, scene.resource_path)
		if rename_error != OK:
			print("Couldn't rename: %s -> %s" % [new_path, scene.resource_path])
			return
		print(" then renamed: %s -> %s" % [new_path, scene.resource_path])


func fix_sprite_frames_resource_file(sprite_frames: SpriteFrames, overwrite_original_file := false) -> void:
	if sprite_frames.resource_path.get_basename().ends_with("_fixed"):
		return # Let's ignore files produced by us.

	if not try_fix_sprite_frames(sprite_frames):
		return # Didn't need fixing.

	# Something was fixed, try to resave.
	var new_path := sprite_frames.resource_path if overwrite_original_file else "%s_fixed.%s" % [sprite_frames.resource_path.get_basename(), sprite_frames.resource_path.get_extension()]
	var save_error := ResourceSaver.save(new_path, sprite_frames, ResourceSaver.FLAG_CHANGE_PATH)
	if save_error != OK:
		print("Couldn't save: %s" % [new_path])
		return

	print("Fixed SpriteFrames: %s" % [sprite_frames.resource_path])
	print("          saved to: %s" % [new_path])

	if overwrite_original_file:
		var rename_error := Directory.new().rename(new_path, sprite_frames.resource_path)
		if rename_error != OK:
			print("Couldn't rename: %s -> %s" % [new_path, sprite_frames.resource_path])
			return
		print(" then renamed: %s -> %s" % [new_path, sprite_frames.resource_path])


func try_fix_sprite_frames(sprite_frames: SpriteFrames) -> bool:
	var anything_fixed := false

	for animation_name in sprite_frames.get_animation_names():
		for frame_index in sprite_frames.get_frame_count(animation_name):
			var frame: Texture = sprite_frames.get_frame(animation_name, frame_index)

			var needs_fixing: bool = frame is AtlasTexture and frame.atlas is AtlasTexture
			if not needs_fixing:
				continue

			var src_atlas: AtlasTexture = frame.atlas
			var src_origin: Vector2 = src_atlas.region.position - src_atlas.margin.position

			var fixed_position: Vector2 = (frame.region.position - frame.margin.position) - src_origin
			var fixed_size: Vector2 = frame.region.size + frame.margin.size

			frame.region = Rect2(fixed_position, fixed_size)
			frame.margin = Rect2()
			anything_fixed = true

	return anything_fixed


func get_sprite_frames_from_node_subtree(node: Node) -> Array:
	return filter_unique_values(_get_sprite_frames_from_node_subtree(node))


func _get_sprite_frames_from_node_subtree(node: Node) -> Array:
	var found_sprite_frames: Array = get_sprite_frames_from_props_of(node)
	for child in node.get_children():
		found_sprite_frames.append_array(_get_sprite_frames_from_node_subtree(child))
	return found_sprite_frames


func get_sprite_frames_from_props_of(obj: Object) -> Array:
	var found_sprite_frames := []
	for property in obj.get_property_list():
		if (property.usage & PROPERTY_USAGE_STORAGE) == 0:
			continue
		if property.hint != PROPERTY_HINT_RESOURCE_TYPE:
			continue
		var sprite_frames = obj.get(property.name) as SpriteFrames
		if sprite_frames:
			found_sprite_frames.push_back(sprite_frames)
	return found_sprite_frames


func filter_unique_values(array: Array) -> Array:
	var dict := {}
	for value in array:
		dict[value] = null
	return dict.keys()
```

</details>

**Comment by kleonc** [comment]
>In retrospect it would be nice to have a fix for this, but I guess I can go through every animated sprite I have and modify every single item. The other problem here is I can't even do a diff because Godot only makes the change when you open a file. Is there not an option/thought to migrate a project fully on opening so people can see all the actual errors rather than having to find them one at a time over a potentially long period of time?

I don't think there is.

> But the changelog does say this, 'We'll update this list with any further regression that the first users of 3.5-stable may report.'
> 
> Could it be updated so anyone else thinking of migrating knows anything using regions will be broken?

Not sure which changelog you're referring to and how it should be updated. cc @akien-mga

**Comment by chucklepie** [comment]
Thanks for the help, but sadly it didn't work. It created 3 _fixed files but there are many more, of the ones it did fix it missed out most frames, e.g.

![image](https://user-images.githubusercontent.com/12863685/211104669-0c5d7377-5f02-412a-bffa-1817d0780d82.png)


And in the output I see this:

Godot Engine v3.5.1.stable.official (c) 2007-2022 Juan Linietsky, Ariel Manzur & Godot Contributors.
--- GDScript language server started ---
Add AutoLoad
Switch Scene Tab
Saving resource --------
[!] Timeline changes detected. Saving: 2818695019
Switch Scene Tab
Delete Resource
Delete Resource
Add Frame
Delete Resource
Add Frame
Switch Scene Tab
Switch Scene Tab
Switch Scene Tab
Switch Scene Tab
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
Couldn't save: _fixed.
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
Couldn't save: _fixed.
Couldn't save: _fixed.
Couldn't save: _fixed.
Couldn't save: _fixed.
Couldn't save: _fixed.
Couldn't save: _fixed.
Couldn't save: _fixed.
Couldn't save: _fixed.
Fixed scene: res://entities/static/VenusFlyTrap.tscn
 saved to: res://entities/static/VenusFlyTrap_fixed.tscn
Fixed scene: res://entities/static/StaticSpore.tscn
 saved to: res://entities/static/StaticSpore_fixed.tscn
Fixed scene: res://entities/static/Turret.tscn
 saved to: res://entities/static/Turret_fixed.tscn
Fixed scene: res://entities/static/TurretSpore.tscn
 saved to: res://entities/static/TurretSpore_fixed.tscn
Couldn't save: _fixed.
Fixed scene: res://entities/environment/DeadPickup.tscn
 saved to: res://entities/environment/DeadPickup_fixed.tscn
Fixed scene: res://entities/particles/Explosion_Main.tscn
 saved to: res://entities/particles/Explosion_Main_fixed.tscn
Fixed scene: res://entities/particles/Explosions.tscn
 saved to: res://entities/particles/Explosions_fixed.tscn
Fixed scene: res://entities/moving/Rex2_Drip.tscn
 saved to: res://entities/moving/Rex2_Drip_fixed.tscn
Couldn't save: _fixed.
Couldn't save: _fixed.
Couldn't save: _fixed.
 Cannot open font file 'res://Fonts/Overlock/Overlock-Black.ttf'.
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://addons/debug-overlay/demo/DemoAutoload.gd:11 - Parse Error: Non-static function "add_monitor" can only be called from an instance.
 res://addons/debug-overlay/demo/DemoAutoload.gd:11 - Parse Error: Non-static function "add_monitor" can only be called from an instance.
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 Resource file not found: res://Achievements/UI/menu_checkpoint_quit.png.
 Resource file not found: res://Achievements/UI/menu_checkpoint_quit.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://Achievements/UI/menu_checkpoint_quit.png
 Resource file not found: res://assets/sprites/environment/room_arrow.png.
 Resource file not found: res://assets/ui/menu_selected.png.
 Resource file not found: res://assets/ui/menu_selected.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/ui/menu_selected.png
 Resource file not found: res://assets/sprites/environment/room_arrow.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/sprites/environment/room_arrow.png
 Resource file not found: res://assets/sprites/environment/room_arrow.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/sprites/environment/room_arrow.png
 Resource file not found: res://assets/ui/menu_selected.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/ui/menu_selected.png
 Resource file not found: res://assets/sprites/environment/room_arrow.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/sprites/environment/room_arrow.png
 Resource file not found: res://assets/sprites/environment/room_arrow.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/sprites/environment/room_arrow.png
 Resource file not found: res://assets/ui/menu_selected.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/ui/menu_selected.png
 Resource file not found: res://assets/sprites/environment/room_arrow.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/sprites/environment/room_arrow.png
 Resource file not found: res://assets/sprites/environment/room_arrow.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/sprites/environment/room_arrow.png
 Resource file not found: res://assets/ui/menu_selected.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/ui/menu_selected.png
 Resource file not found: res://assets/ui/menu_selected.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/ui/menu_selected.png
 Resource file not found: res://assets/sprites/environment/room_arrow.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/sprites/environment/room_arrow.png
 Resource file not found: res://assets/sprites/environment/room_arrow.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/sprites/environment/room_arrow.png
 Resource file not found: res://assets/ui/menu_selected.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/ui/menu_selected.png
 Resource file not found: res://assets/ui/menu_selected.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/ui/menu_selected.png
 Resource file not found: res://assets/sprites/environment/room_arrow.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/sprites/environment/room_arrow.png
 Resource file not found: res://assets/sprites/environment/room_arrow.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/sprites/environment/room_arrow.png
 Resource file not found: res://assets/ui/menu_selected.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/ui/menu_selected.png
 Resource file not found: res://assets/ui/menu_selected.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/ui/menu_selected.png
 Resource file not found: res://assets/sprites/environment/room_arrow.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/sprites/environment/room_arrow.png
 Resource file not found: res://assets/sprites/environment/room_arrow.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://assets/sprites/environment/room_arrow.png
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 Resource file not found: res://dialogue/character-squidgy.png.
 Resource file not found: res://dialogue/character-squidgy.png.
 scene/resources/resource_format_text.cpp:157 - Couldn't load external resource: res://dialogue/character-squidgy.png
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:155 - Invalid cast: can't convert a non-object value to an object type.
 res://SpriteFrames_fixer.gd:142 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:144 - Invalid type in function 'append_array' in base 'Array'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:138 - Invalid type in function 'filter_unique_values' in base 'EditorScript (SpriteFrames_fixer.gd)'. Cannot convert argument 1 from Nil to Array.
 res://SpriteFrames_fixer.gd:51 - Trying to assign value of type 'Nil' to a variable of type 'Array'.
Switch Scene Tab
Switch Scene Tab
 Signal 'area_entered' is already connected to given method '_on_Hurtbox_area_entered' in that object.
 Signal 'timeout' is already connected to given method '_on_HurtTimer_timeout' in that object.
Switch Scene Tab
 Signal 'area_entered' is already connected to given method '_on_Hurtbox_area_entered' in that object.
 Signal 'timeout' is already connected to given method '_on_HurtTimer_timeout' in that object.
Switch Scene Tab
Switch Scene Tab
Switch Scene Tab

**Comment by chucklepie** [comment]
To save you the trouble, I'll just convert everything by hand.

**Comment by dthomason** [comment]
I just discovered this bug when upgrading from 4.0 beta 1 to beta 10.  So something occurred between then that caused this.  I think they're related. First noticed the position got changed for a couple body parts,  Then when I go into "edit region" It's blank.  Also when I attempt to move anything it glitches to the point of being unable to actually move the sprite
<img width="1356" alt="Screen Shot 2023-01-22 at 9 23 00 PM" src="https://user-images.githubusercontent.com/8356638/213960772-8b9148bc-0c54-4e94-ad8c-8f4654b7a62e.png">

**Comment by kleonc** [comment]
> I just discovered this bug when upgrading from 4.0 beta 1 to beta 10. So something occurred between then that caused this. I think they're related. First noticed the position got changed for a couple body parts, Then when I go into "edit region" It's blank. Also when I attempt to move anything it glitches to the point of being unable to actually move the sprite

I'd say the cause must be different than for what is reported in here by @chucklepie. #56796 was merged even before `4.0.alpha1` so your `4.0.beta1` project was already created with the relevant changes (and it was working I guess). Can't really tell what's the issue for your case without more details. Please open a separate issue for it, and provide a minimal reproduction project (could be even two projects - before and after upgrading; if a single project then pre-upgrading would be better as I could upgrade myself to see the result). Feel free to ping me (@kleonc) in there.
